--- elf/elf.h	2005-04-05 18:24:53.000000000 -0400
+++ elf/elf.h	2005-04-05 18:36:08.000000000 -0400
@@ -1761,6 +1761,29 @@
 #define R_PARISC_LTOFF_TP16F	229	/* 16 bits LT-TP-rel. address.  */
 #define R_PARISC_LTOFF_TP16WF	230	/* 16 bits LT-TP-rel. address.  */
 #define R_PARISC_LTOFF_TP16DF	231	/* 16 bits LT-TP-rel. address.  */
+
+#define R_PARISC_GNU_VTENTRY    232
+#define R_PARISC_GNU_VTINHERIT  233
+#define R_PARISC_TLS_GD21L      234
+#define R_PARISC_TLS_GD14R      235
+#define R_PARISC_TLS_GDCALL     236
+#define R_PARISC_TLS_LDM21L     237
+#define R_PARISC_TLS_LDM14R     238
+#define R_PARISC_TLS_LDMCALL    239
+#define R_PARISC_TLS_LDO21L     240
+#define R_PARISC_TLS_LDO14R     241
+#define R_PARISC_TLS_DTPMOD32   242	/* FIXME: Verify support in dl-machine */
+#define R_PARISC_TLS_DTPMOD64   243
+#define R_PARISC_TLS_DTPOFF32   244	/* FIXME: Likewise */
+#define R_PARISC_TLS_DTPOFF64   245
+
+#define R_PARISC_TLS_LE21L      R_PARISC_TPREL21L
+#define R_PARISC_TLS_LE14R      R_PARISC_TPREL14R
+#define R_PARISC_TLS_IE21L      R_PARISC_LTOFF_TP21L
+#define R_PARISC_TLS_IE14R      R_PARISC_LTOFF_TP14R
+#define R_PARISC_TLS_TPREL32    R_PARISC_TPREL32	/* FIXME: Likewise*/
+#define R_PARISC_TLS_TPREL64    R_PARISC_TPREL64
+
 #define R_PARISC_HIRESERVE	255
 
 /* Legal values for p_type field of Elf32_Phdr/Elf64_Phdr.  */
