--- libc.orig/stdio-common/vfscanf.c	2005-07-18 06:10:03.000000000 +0200
+++ libc/stdio-common/vfscanf.c	2007-08-17 00:38:50.000000000 +0200
@@ -547,12 +547,17 @@
 	{
 	  /* Eat whitespace.  */
 	  int save_errno = errno;
-	  errno = 0;
+	  __set_errno (0);
 	  do
-	    if (inchar () == EOF && errno == EINTR)
+	    /* We add the additional test for EOF here since otherwise
+	       inchar will restore the old errno value which might be
+	       EINTR but does not indicate an interrupt since nothing
+	       was read at this time.  */
+	    if (__builtin_expect ((c == EOF || inchar () == EOF)
+				  && errno == EINTR, 0))
 	      input_error ();
 	  while (ISSPACE (c));
-	  errno = save_errno;
+	  __set_errno (save_errno);
 	  ungetc (c, s);
 	  skip_space = 0;
 	}
