--- linuxthreads/sysdeps/pthread/gai_misc.h.orig	2007-02-05 07:51:33.000000000 +0100
+++ linuxthreads/sysdeps/pthread/gai_misc.h	2007-02-05 07:55:01.000000000 +0100
@@ -34,7 +34,12 @@
 {
   sigset_t ss;
   sigemptyset (&ss);
+#ifdef __linux__
+  INTERNAL_SYSCALL_DECL (err);
+  INTERNAL_SYSCALL (rt_sigprocmask, err, 4, SIG_SETMASK, &ss, NULL, _NSIG / 8);
+#else
   __sigprocmask (SIG_SETMASK, &ss, NULL);
+#endif
 }
 
 extern inline int
@@ -55,12 +60,22 @@
   sigset_t ss;
   sigset_t oss;
   sigfillset (&ss);
+#ifdef __linux__
+  INTERNAL_SYSCALL_DECL (err);
+  INTERNAL_SYSCALL (rt_sigprocmask, err, 4, SIG_SETMASK, &ss, &oss, _NSIG / 8);
+#else
   __sigprocmask (SIG_SETMASK, &ss, &oss);
+#endif
 
   int ret = pthread_create (threadp, &attr, tf, arg);
 
   /* Restore the signal mask.  */
+#ifdef __linux__
+  INTERNAL_SYSCALL (rt_sigprocmask, err, 4, SIG_SETMASK, &oss, NULL,
+		    _NSIG / 8);
+#else
   __sigprocmask (SIG_SETMASK, &oss, NULL);
+#endif
 
   (void) pthread_attr_destroy (&attr);
   return ret;
