Index: glibc-2.5/argp/argp.h
===================================================================
--- glibc-2.5.orig/argp/argp.h
+++ glibc-2.5/argp/argp.h
@@ -33,13 +33,36 @@
 # define __const const
 #endif
 
-#ifndef __THROW
-# define __THROW
+#ifdef __THROW
+# undef __THROW
 #endif
-#ifndef __NTH
-# define __NTH(fct) fct __THROW
+#ifdef __NTH
+# undef __NTH
 #endif
 
+#ifdef __GNUC__
+
+/* see <sys/cdefs.h> */
+# if !defined __cplusplus && __GNUC_PREREQ (3, 3)
+#  define __THROW	__attribute__ ((__nothrow__))
+#  define __NTH(fct)	__attribute__ ((__nothrow__)) fct
+# else
+#  if defined __cplusplus && __GNUC_PREREQ (2,8)
+#   define __THROW	throw ()
+#   define __NTH(fct)	fct throw ()
+#  else
+#   define __THROW
+#   define __NTH(fct)	fct
+#  endif
+# endif
+
+#else	/* Not GCC.  */
+
+# define __THROW
+# define __NTH(fct)	fct
+
+#endif	/* GCC.  */
+
 #ifndef __attribute__
 /* This feature is available in gcc versions 2.5 and later.  */
 # if __GNUC__ < 2 || (__GNUC__ == 2 && __GNUC_MINOR__ < 5) || __STRICT_ANSI__
