commit e2930d8777c0331b5882e27cbb53f3cf8516a62d
Author: Samuel Thibault <samuel.thibault@ens-lyon.org>
Date:   Tue Aug 31 14:36:08 2021 +0200

    hurd: Fix typo in msync
    
    == has higher priority than &

diff --git a/sysdeps/mach/hurd/msync.c b/sysdeps/mach/hurd/msync.c
index 4311dd56d1..a1244e42e2 100644
--- a/sysdeps/mach/hurd/msync.c
+++ b/sysdeps/mach/hurd/msync.c
@@ -47,7 +47,7 @@ msync (void *addr, size_t length, int flags)
 
   kern_return_t err;
 
-  if (flags & (MS_SYNC | MS_ASYNC) == (MS_SYNC | MS_ASYNC))
+  if ((flags & (MS_SYNC | MS_ASYNC)) == (MS_SYNC | MS_ASYNC))
     return __hurd_fail (EINVAL);
 
   while (cur < target)
