http://sourceware.org/bugzilla/show_bug.cgi?id=3992

Index: stdlib/fmtmsg.c
===================================================================
RCS file: /cvs/glibc/libc/stdlib/fmtmsg.c,v
retrieving revision 1.20
diff -u -d -p -r1.20 stdlib/fmtmsg.c
--- stdlib/fmtmsg.c	15 May 2006 18:41:18 -0000	1.20
+++ stdlib/fmtmsg.c	7 Feb 2007 14:06:37 -0000
@@ -20,6 +20,7 @@
 #include <fmtmsg.h>
 #include <bits/libc-lock.h>
 #include <stdio.h>
+#include <stdint.h>
 #include <stdlib.h>
 #include <string.h>
 #include <sys/syslog.h>
