commit 6b2fca45129f07f1575a5a6e5a719090b59555d6
Author: Richard Braun <rbraun@sceen.net>
Date:   Wed Sep 5 11:27:05 2012 +0200

    Hurd: fix sigstate_is_global_rcv

diff --git a/hurd/hurdsig.c b/hurd/hurdsig.c
index 736ce3d..87c685b 100644
--- a/hurd/hurdsig.c
+++ b/hurd/hurdsig.c
@@ -159,7 +159,8 @@ _hurd_sigstate_set_global_rcv (struct hurd_sigstate *ss)
 static int
 sigstate_is_global_rcv (const struct hurd_sigstate *ss)
 {
-  return ss->actions[0].sa_handler == SIG_IGN;
+  return (_hurd_global_sigstate != NULL)
+	 && (ss->actions[0].sa_handler == SIG_IGN);
 }
 
 /* Lock/unlock a hurd_sigstate structure.  If the accessors below require
