Index: glibc-2.7/localedata/locales/am_ET
===================================================================
--- glibc-2.7.orig/localedata/locales/am_ET	2007-10-23 08:48:53.000000000 -0400
+++ glibc-2.7/localedata/locales/am_ET	2007-10-23 08:48:58.000000000 -0400
@@ -62,7 +62,7 @@
 %
 %  The ETHIOPIC script declaration must come first.
 %
-script <ETHIOPIC>
+define USE_ETHIOPIC
 %
 %  Correcting Unicode's linguistic ordering to the traditional.
 %
Index: glibc-2.7/localedata/locales/iso14651_t1_common
===================================================================
--- glibc-2.7.orig/localedata/locales/iso14651_t1_common	2007-10-23 08:46:16.000000000 -0400
+++ glibc-2.7/localedata/locales/iso14651_t1_common	2007-10-23 08:48:58.000000000 -0400
@@ -50,6 +50,9 @@
 # 2005-11-29, Pablo Saratxaga <pablo@mandriva.com>
 
 # Déclaration des systèmes d'écriture / Declaration of scripts
+ifdef USE_ETHIOPIC
+script <ETHIOPIC>
+endif
 script <SPECIAL>
 script <LATIN>
 script <TIFINAGH>
