Index: glibc-2.7/localedata/locales/fr_CA
===================================================================
--- glibc-2.7.orig/localedata/locales/fr_CA	2007-10-22 22:47:54.000000000 -0400
+++ glibc-2.7/localedata/locales/fr_CA	2007-10-22 22:48:09.000000000 -0400
@@ -121,6 +121,7 @@
 date_fmt       "<U0025><U0061><U0020><U0025><U0062><U0020><U0025><U0065>/
 <U0020><U0025><U0048><U003A><U0025><U004D><U003A><U0025><U0053><U0020>/
 <U0025><U005A><U0020><U0025><U0059>"
+first_weekday 1
 END LC_TIME
 
 LC_PAPER
