Index: glibc-2.6/locale/programs/ld-collate.c
===================================================================
--- glibc-2.6.orig/locale/programs/ld-collate.c
+++ glibc-2.6/locale/programs/ld-collate.c
@@ -2808,8 +2808,18 @@
 	  if (state == 1 || state == 3 || state == 5)
 	    goto err_label;
 
-	  load_locale (LC_COLLATE, now->val.str.startmb, repertoire_name,
-		       charmap, result);
+          /* Get the locale definition.  */
+          copy_locale = load_locale (LC_COLLATE, now->val.str.startmb,
+                                     repertoire_name, charmap, NULL);
+          if ((copy_locale->avail & COLLATE_LOCALE) == 0)
+            {
+              /* Not yet loaded.  So do it now.  */
+              if (locfile_read (copy_locale, charmap) != 0)
+                goto skip_category;
+            }
+
+          if (copy_locale->categories[LC_COLLATE].collate == NULL)
+            return;
 
 	  lr_ignore_rest (ldfile, 1);
 	  break;
